openwisp_controller/subnet_division/rule_types/base.py

```diff
@@ -0,0 +1,172 @@
+import logging
+from ipaddress import ip_network
+from operator import attrgetter
+
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import transaction
+from django.dispatch import Signal
+from django.utils.translation import ugettext_lazy as _
+from netaddr import IPNetwork
+from swapper import load_model
+
+logger = logging.getLogger(__name__)
+
+Subnet = load_model('openwisp_ipam', 'Subnet')
+IpAddress = load_model('openwisp_ipam', 'IpAddress')
+SubnetDivisionRule = load_model('subnet_division', 'SubnetDivisionRule')
+SubnetDivisionIndex = load_model('subnet_division', 'SubnetDivisionIndex')
+VpnClient = load_model('config', 'VpnClient')
+
+
+class BaseSubnetDivisionRuleType(object):
+    provision_signal = None
+    provision_sender = None
+    provision_dispatch_uid = None
+
+    destroyer_signal = None
+    destroyer_sender = None
+    destroyer_dispatch_uid = None
+
+    organization_path = None
+    subnet_path = None
+
+    @classmethod
+    def validate_rule_type(cls):
+        assert issubclass(cls, BaseSubnetDivisionRuleType)
+
+        assert isinstance(cls.provision_signal, Signal)
+        assert isinstance(cls.provision_dispatch_uid, str)
+        cls.provision_sender = load_model(*cls.provision_sender)
+
+        assert isinstance(cls.destroyer_signal, Signal)
+        assert isinstance(cls.destroyer_dispatch_uid, str)
+        cls.destroyer_sender = load_model(*cls.destroyer_sender)
+
+        assert isinstance(cls.organization_path, str)
+        assert isinstance(cls.subnet_path, str)
+
+    @classmethod
+    def provision_receiver(cls, instance, **kwargs):
+        rule_type = f'{cls.__module__}.{cls.__name__}'
+        transaction.on_commit(
+            lambda: cls.create_subnets_ips(instance, rule_type, **kwargs)
+        )
+
+    @classmethod
+    def destroyer_receiver(cls, instance, **kwargs):
+        from ..handlers import destroy_provisioned_subnets
```

Comment:
why this?

Why not just putting the code here? I don't see `destroy_provisioned_subnets` being called anywhere else, I think the logic belongs to this class.